File: apps/tg_bot/lib/tg_bot/messenger.ex
defmodule TGBot.Messenger do

  alias Nadia.Model.InlineKeyboardButton
  alias Nadia.Model.InlineKeyboardMarkup

  @spec send_text(integer, String.t, Keyword.t) :: integer
  def send_text(chat_id, text, opts \\ []) do
    opts = transform_opts(opts)
    send_message(chat_id, text, opts)
  end

  @spec send_markdown(integer, String.t, Keyword.t) :: integer
  def send_markdown(chat_id, text, opts \\ []) do
    opts = transform_opts(opts)
    send_message(
      chat_id,
      text,
      opts ++
      [
        parse_mode: "Markdown",
        disable_web_page_preview: true
      ]
    )
  end

  @spec send_photo(integer, binary, Keyword.t) :: integer
  def send_photo(chat_id, photo, opts \\ []) do
    opts = transform_opts(opts)
    case Nadia.send_photo(chat_id, photo, opts) do
      {:error, error} -> raise error
      {:ok, msg} -> msg.message_id
    end
  end

  @spec send_notification(String.t, String.t) :: any
  def send_notification(callback_id, text) do
    answer_callback(callback_id, text: text)
  end

  @spec answer_callback(String.t, Keyword.t) :: any
  def answer_callback(callback_id, opts \\ []) do
    case Nadia.answer_callback_query(callback_id, opts) do
      {:error, error} -> raise error
      _ -> nil
    end
  end

  @spec delete_keyboard(integer, integer) :: integer
  def delete_keyboard(chat_id, message_id) do
    case Nadia.API.request("editMessageReplyMarkup", chat_id: chat_id, message_id: message_id) do
      {:error, error} -> raise error
      {:ok, msg} -> msg.message_id
    end
  end

  @spec send_photos(integer, [map()]) :: any
  def send_photos(chat_id, photos) do
    media = Enum.map(
      photos,
      fn (photo) -> %{type: "photo", media: photo.url, caption: photo.caption} end
    )
    media_encoded = Poison.encode!(media)
    try do
      case Nadia.API.request("sendMediaGroup", chat_id: chat_id, media: media_encoded) do
        {:error, error} -> raise error
        _ -> nil
      end
    rescue
      FunctionClauseError -> nil
    end
  end

  @spec transform_opts(Keyword.t) :: Keyword.t
  defp transform_opts(opts) do
    {keyboard, opts} = Keyword.pop(opts, :keyboard)
    reply_markup = transform_keyboard(keyboard)
    if reply_markup, do: Keyword.put(opts, :reply_markup, reply_markup), else: opts
  end

  defp transform_keyboard(nil), do: nil
  @spec transform_keyboard([[map()]]) :: InlineKeyboardMarkup.t
  defp transform_keyboard(keyboard_data) do
    keyboard = Enum.map(
      keyboard_data,
      fn keyboard_line ->
        Enum.map(
          keyboard_line,
          fn item_data ->
            %InlineKeyboardButton{
              text: item_data.text,
              callback_data: item_data.payload,
              url: "",
              switch_inline_query: ""
            }
          end
        )
      end
    )
    %InlineKeyboardMarkup{inline_keyboard: keyboard}
  end

  @spec send_message(integer, String.t, Keyword.t) :: integer
  defp send_message(chat_id, text, opts \\ []) do
    case Nadia.send_message(chat_id, text, opts) do
      {:error, error} -> raise error
      {:ok, msg} -> msg.message_id
    end
  end
end

File: tg_bot/messenger/messenger.go
package messenger

import (
	"bytes"
	"context"
	"encoding/json"
	"github.com/gazoon/go-utils/logging"
	"github.com/pkg/errors"
	"github.com/satori/go.uuid"
	"gopkg.in/telegram-bot-api.v4"
	"io"
	"mime/multipart"
	"net/http"
	"net/url"
	"strconv"
	"time"
)

var (
	httpTimeout = time.Second * 10
	apiURL      = "https://api.telegram.org/bot"
)

const (
	sendPhotoMethod = "sendPhoto"
)

type messageResponse struct {
	Ok          bool              `json:"ok"`
	Result      *tgbotapi.Message `json:"result"`
	ErrorCode   int               `json:"error_code"`
	Description string            `json:"description"`
}

type Telegram struct {
	*logging.LoggerMixin
	botAPI *tgbotapi.BotAPI
	token  string
	client *http.Client
}

func NewTelegram(token string) (*Telegram, error) {
	httpClient := &http.Client{Timeout: httpTimeout}
	bot, err := tgbotapi.NewBotAPIWithClient(token, httpClient)
	if err != nil {
		return nil, errors.Wrap(err, "Telegram API initialization")
	}
	return &Telegram{
		botAPI:      bot,
		token:       token,
		client:      httpClient,
		LoggerMixin: logging.NewLoggerMixin("messenger", nil),
	}, nil
}

func (self *Telegram) SendText(ctx context.Context, chatId int, text string,
	opts ...func(msg *tgbotapi.MessageConfig)) (int, error) {
	msg := tgbotapi.NewMessage(int64(chatId), text)
	for _, o := range opts {
		o(&msg)
	}
	return self.send(msg)
}

func (self *Telegram) SendMarkdown(ctx context.Context, chatId int, text string,
	opts ...func(msg *tgbotapi.MessageConfig)) (int, error) {
	opts = append(opts, func(msg *tgbotapi.MessageConfig) {
		msg.ParseMode = "Markdown"
		msg.DisableWebPagePreview = true
	})
	return self.SendText(ctx, chatId, text, opts...)
}

func (self *Telegram) GetMembersNum(ctx context.Context, chatId int) (int, error) {
	c := tgbotapi.ChatConfig{ChatID: int64(chatId)}
	num, err := self.botAPI.GetChatMembersCount(c)
	return num, errors.Wrap(err, "get chat members tg API call")
}

func (self *Telegram) SendNotification(ctx context.Context, callbackId, text string) error {
	_, err := self.botAPI.AnswerCallbackQuery(
		tgbotapi.CallbackConfig{CallbackQueryID: callbackId, Text: text})
	return errors.Wrap(err, "answer callback tg API call, with text")
}

func (self *Telegram) AnswerCallback(ctx context.Context, callbackId string) error {
	_, err := self.botAPI.AnswerCallbackQuery(
		tgbotapi.CallbackConfig{CallbackQueryID: callbackId})
	return errors.Wrap(err, "answer callback tg API call")
}

func (self *Telegram) SendBinaryPhoto(ctx context.Context, chatId int,
	fileData io.Reader, opts ...func(settings *tgbotapi.PhotoConfig)) (int, string, error) {
	c := tgbotapi.PhotoConfig{}
	for _, o := range opts {
		o(&c)
	}
	body := &bytes.Buffer{}
	writer := multipart.NewWriter(body)
	part, err := writer.CreateFormFile("photo", uuid.NewV4().String())
	if err != nil {
		return 0, "", errors.Wrap(err, "create form for file uploading")
	}
	_, err = io.Copy(part, fileData)
	writer.WriteField("chat_id", strconv.Itoa(chatId))
	if c.Caption != "" {
		writer.WriteField("caption", c.Caption)
	}
	if c.ReplyMarkup != nil {
		replayMarkupData, err := json.Marshal(c.ReplyMarkup)
		if err != nil {
			return 0, "", errors.Wrap(err, "serialize replay markup into json")
		}

		writer.WriteField("reply_markup", string(replayMarkupData))
	}
	err = writer.Close()
	if err != nil {
		return 0, "", errors.Wrap(err, "close multipart writer")
	}

	resp, err := self.client.Post(self.buildUrl("sendPhoto"), writer.FormDataContentType(), body)
	return handleSendPhotoResponse(resp, err)
}

func (self *Telegram) SendPhoto(ctx context.Context, chatId int,
	photoUri string, opts ...func(settings *tgbotapi.PhotoConfig)) (int, string, error) {

	c := tgbotapi.NewPhotoUpload(int64(chatId), photoUri)
	for _, o := range opts {
		o(&c)
	}
	data := url.Values{}
	data.Set("chat_id", strconv.Itoa(chatId))
	data.Set("photo", photoUri)
	if c.Caption != "" {
		data.Set("caption", c.Caption)
	}
	if c.ReplyMarkup != nil {
		replayMarkupData, err := json.Marshal(c.ReplyMarkup)
		if err != nil {
			return 0, "", errors.Wrap(err, "serialize replay markup into json")
		}
		data.Set("reply_markup", string(replayMarkupData))
	}

	self.GetLogger(ctx).WithField("photo_uri", photoUri).Info("Upload photo")
	resp, err := self.client.PostForm(self.buildUrl(sendPhotoMethod), data)
	return handleSendPhotoResponse(resp, err)
}

func (self *Telegram) DeleteAttachedKeyboard(ctx context.Context, chatId,
	messageId int) (int, error) {
	c := tgbotapi.EditMessageReplyMarkupConfig{
		BaseEdit: tgbotapi.BaseEdit{ChatID: int64(chatId), MessageID: messageId},
	}
	return self.send(c)
}

func (self *Telegram) send(c tgbotapi.Chattable) (int, error) {
	resp, err := self.botAPI.Send(c)
	if err != nil {
		return 0, errors.Wrapf(err, "send %T tg API call", c)
	}
	return resp.MessageID, nil
}

func (self *Telegram) buildUrl(methodName string) string {
	return apiURL + self.token + "/" + methodName
}

func handleSendPhotoResponse(resp *http.Response, err error) (int, string, error) {
	if err != nil {
		return 0, "", errors.Wrap(err, "send multipart post request")
	}
	defer resp.Body.Close()
	apiResponse := &messageResponse{}
	err = json.NewDecoder(resp.Body).Decode(apiResponse)
	if err != nil {
		return 0, "", errors.Wrap(err, "parse sendPhoto response into json")
	}
	if !apiResponse.Ok || apiResponse.Result == nil {
		return 0, "", errors.Errorf("invalid sendPhoto response: %s", apiResponse.Description)
	}
	return retrievePhotoInfo(apiResponse.Result)

}

func retrievePhotoInfo(resp *tgbotapi.Message) (int, string, error) {
	if resp.Photo == nil || len(*resp.Photo) == 0 {
		return 0, "", errors.New("send photo response without photos info")
	}
	fileId := (*resp.Photo)[len(*resp.Photo)-1].FileID
	return resp.MessageID, fileId, nil

}
